server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/ClusterBalanceStats.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.routing.allocation.allocator;
+
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.cluster.node.DiscoveryNodeRole;
+import org.elasticsearch.cluster.routing.RoutingNode;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.allocation.WriteLoadForecaster;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.util.Maps;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.ToDoubleFunction;
+
+import static java.util.stream.Collectors.summarizingDouble;
+
+public record ClusterBalanceStats(Map<String, TierBalanceStats> tiers) implements Writeable, ToXContentObject {
+
+    public static ClusterBalanceStats EMPTY = new ClusterBalanceStats(Map.of());
+
+    public static ClusterBalanceStats createFrom(ClusterState clusterState, WriteLoadForecaster writeLoadForecaster) {
+        var tierToNodeStats = new HashMap<String, List<NodeStats>>();
```

Comment:
I was wondering if can calculate the stats in one go with the Streams API using something like that
```java
var tierToNodeStats = clusterState.getRoutingNodes()
    .stream()
    .flatMap(
        routingNode -> routingNode.node()
            .getRoles()
            .stream()
            .filter(DiscoveryNodeRole::canContainData)
            .filter(role -> Objects.equals(DiscoveryNodeRole.DATA_ROLE, role) == false)
            .map(role -> Map.entry(role, NodeStats.createFrom(routingNode, clusterState.metadata(), writeLoadForecaster)))
    )
    .collect(groupingBy(e -> e.getKey().roleName(), mapping(Map.Entry::getValue, toList())));
```

Reply:
I do not think that this is very important, just wanted to avoid computing counts for a node more then once in case multiple roles are defined.